internal/cache/memcache/memory_cache.go

```diff
@@ -0,0 +1,128 @@
+// A simple memory-only thread-safe cache with TTL.
+package memcache
+
+import (
+	"runtime"
+	"sync"
+	"time"
+)
+
+// Cache stores arbitrary data with expiration time.
+type Cache[K comparable, V any] struct {
+	items   map[K]*item[V]
+	mu      sync.Mutex
+	done    chan any
+	clean   chan bool
+	once    sync.Once
+	cleanWG sync.WaitGroup
+}
+
+// An item represents arbitrary data with expiration time.
+type item[V any] struct {
+	data    V
+	expires int64
+}
+
+// New creates a new cache that asynchronously cleans
+// expired entries after the given time passes. If cleaningInterval
+// is zero, no background cleanup goroutine is scheduled.
+func New[K comparable, V any](cleaningInterval time.Duration) *Cache[K, V] {
+	cache := &Cache[K, V]{
+		items: make(map[K]*item[V]),
+		clean: make(chan bool),
+		done:  make(chan any),
+	}
+
+	if cleaningInterval != 0 {
+		go func() {
+			ticker := time.NewTicker(cleaningInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					cache.cleanup()
+				case <-cache.clean:
+					cache.cleanup()
+					cache.cleanWG.Done()
+				case <-cache.done:
+					return
+				}
+			}
+		}()
+	}
+
+	// Shutdown the goroutine when GC wants to clean this up
+	runtime.SetFinalizer(cache, func(c *Cache[K, V]) {
+		c.Stop()
+	})
+
+	return cache
+}
+
+// cleanup function is called from the background goroutine
+func (cache *Cache[K, V]) cleanup() {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	now := time.Now().UnixNano()
+	for key, item := range cache.items {
+		if item.expires > 0 && now > item.expires {
+			delete(cache.items, key)
+		}
+	}
+}
+
+// Get gets the value for the given key.
+func (cache *Cache[K, V]) Get(key K) (V, bool) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	item, exists := cache.items[key]
+	if !exists || (item.expires > 0 && time.Now().UnixNano() > item.expires) {
+		var nothing V
+		return nothing, false
+	}
```

Comment:
TIL: This is how you return default value for a generic type (e.g. `0` for `int` or `nil` for `struct`).